### PR TITLE
Feature/sdl 0273 webengine projection mode

### PR DESCRIPTION
**Pull request by LuxoftAKutsan**
Implements  [#375](https://github.com/smartdevicelink/sdl_hmi/issues/375)

This PR is **ready** for review.

### Testing Plan
 - [x] Manual Testing 

### Summary
HMI implementation for manual testing of the proposal : https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0273-webengine-projection-mode.md. https://github.com/smartdevicelink/sdl_evolution/issues/912 

Manual instruction: [Manual instruction for [SDL-0273] WebEngine Projection mode .xlsx](https://github.com/smartdevicelink/sdl_hmi/files/4889427/Manual.instruction.for.SDL-0273.WebEngine.Projection.mode.xlsx)

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: app/controller/sdl/Abstract/Controller.js [hunk 1]
```diff
@@ -1495,13 +1526,50 @@ SDL.SDLController = Em.Object.extend(
...
+          if (template == "DEFAULT" && model.appType.indexOf('WEB_VIEW') >= 0) {
+            return true;
+          }
+
+          return template == "WEB_VIEW";
+        }
+
+        var appModel = this.getApplicationModel(appID);
+        if (appModel && is_web_view_template(appModel) && windowType == "MAIN") {
+          let text_fields = windowCapability["systemCapability"]["displayCapabilities"][0]["windowCapabilities"][0]["textFields"];
```
Comment (kshala-ford) [inline app/controller/sdl/Abstract/Controller.js:1546]: I'm not an expert in JS but is `text_fields` referring to the `windowCapabilities` or is this "let" creating a copy? Are any changes to `text_fields` reflected to `windowCapability["systemCapability"]["displayCapabilities"][0]["windowCapabilities"][0]["textFields"]`?

File: app/controller/sdl/Abstract/Controller.js [hunk 2]
```diff
@@ -1495,13 +1526,50 @@ SDL.SDLController = Em.Object.extend(
...
+            'mainField1', 'mainField2', 'mainField3', 'mainField4',
+            'mediaClock', 'mediaTrack'
+          ];
+          for (var i = text_fields.length - 1; i >= 0; i--) {
+            if (text_fields_to_exclude.includes(text_fields[i].name)) {
+              text_fields.splice(i, 1);
+            }
+          }
+
+          let image_fields = windowCapability["systemCapability"]["displayCapabilities"][0]["windowCapabilities"][0]["imageFields"];
```
Comment (kshala-ford) [inline app/controller/sdl/Abstract/Controller.js:1557]: Same question as to `image_fields`. Is this variable referring to the same array as `windowCapability["systemCapability"]["displayCapabilities"][0]["windowCapabilities"][0]["imageFields"]` or to a copy of it?

File: app/controller/sdl/Abstract/Controller.js [hunk 3]
```diff
@@ -1495,13 +1526,50 @@ SDL.SDLController = Em.Object.extend(
...
+          const text_fields_to_exclude = [
+            'mainField1', 'mainField2', 'mainField3', 'mainField4',
+            'mediaClock', 'mediaTrack'
+          ];
+          for (var i = text_fields.length - 1; i >= 0; i--) {
+            if (text_fields_to_exclude.includes(text_fields[i].name)) {
+              text_fields.splice(i, 1);
+            }
+          }
+
```
Comment (kshala-ford) [inline app/controller/sdl/Abstract/Controller.js:1556]: I believe you will need to write `text_fields` back to `windowCapability["systemCapability"]["displayCapabilities"][0]["windowCapabilities"][0]["textFields"]` but I might be wrong. Please let me know if this is needed.

File: app/controller/sdl/Abstract/Controller.js [hunk 4]
```diff
@@ -1495,13 +1526,50 @@ SDL.SDLController = Em.Object.extend(
...
+          let image_fields = windowCapability["systemCapability"]["displayCapabilities"][0]["windowCapabilities"][0]["imageFields"];
+          const image_fields_to_exclude = [
+            'softButtonImage', 'menuIcon', 'graphic', 'secondaryGraphic'
+          ];
+          for (var i = image_fields.length - 1; i >= 0; i--) {
+            if (image_fields_to_exclude.includes(image_fields[i].name)) {
+              image_fields.splice(i, 1);
+            }
+          }
+
```
Comment (kshala-ford) [inline app/controller/sdl/Abstract/Controller.js:1566]: I believe you will need to write `image_fields` back to `windowCapability["systemCapability"]["displayCapabilities"][0]["windowCapabilities"][0]["imageFields"]` but I might be wrong. Please let me know if this is needed.

File: app/model/sdl/Abstract/data.js [hunk 5]
```diff
@@ -1021,7 +1025,7 @@ SDL.SDLModelData = Em.Object.create(
               ],
               "imageTypeSupported": ["STATIC", "DYNAMIC"],
               "numCustomPresetsAvailable": 8,
-              "templatesAvailable": ["TEXT_WITH_GRAPHIC", "BUTTONS_WITH_GRAPHIC", "GRAPHIC_WITH_TEXT"],
+              "templatesAvailable": ["MEDIA", "NON-MEDIA", "DEFAULT", "NAV_FULLSCREEN_MAP", 'WEB_VIEW'],
```
Comment (kshala-ford) [inline app/model/sdl/Abstract/data.js:1028]: I understand adding `WEB_VIEW` but why are all templates available changed here?

File: app/controller/sdl/Abstract/Controller.js (hunk 1 above)
Comment (AKalinich-Luxoft) [reply app/controller/sdl/Abstract/Controller.js:1546]: @kshala-ford not exactly. `let` just declares a new variable in ES6 standard. But please take a look at line 1529 - there are a `SDL.deepCopy` method. This method is doing exactly what you think - creates a copy of initial window capabilities. So `windowCapability` variable is already contains a local copy so we can safely modify it as we want.
It does not affect source capabilities stored in a separate resource file.

Comment (AKalinich-Luxoft) [reply app/controller/sdl/Abstract/Controller.js:1546]: @kshala-ford `text_fields` at this line of code is a reference to the same array, its not a copy. However, as I mentioned above, `windowCapabilities` is a copy so we are OK with modifying its structure

File: app/controller/sdl/Abstract/Controller.js (hunk 2 above)
Comment (AKalinich-Luxoft) [reply app/controller/sdl/Abstract/Controller.js:1557]: @kshala-ford yes, the same as above

File: app/controller/sdl/Abstract/Controller.js (hunk 3 above)
Comment (AKalinich-Luxoft) [reply app/controller/sdl/Abstract/Controller.js:1556]: @kshala-ford no, as `image_fields` is a reference to `windowCapability["systemCapability"]["displayCapabilities"][0]["windowCapabilities"][0]["imageFields"]` so we are making changes directly, using `image_fields` and its methods

File: app/controller/sdl/Abstract/Controller.js (hunk 4 above)
Comment (AKalinich-Luxoft) [reply app/controller/sdl/Abstract/Controller.js:1566]: @kshala-ford no, as `image_fields` is a reference to `windowCapability["systemCapability"]["displayCapabilities"][0]["windowCapabilities"][0]["imageFields"]` so we are making changes directly, using `image_fields` and its methods

File: app/model/sdl/Abstract/data.js (hunk 5 above)
Comment (AKalinich-Luxoft) [reply app/model/sdl/Abstract/data.js:1028]: @kshala-ford as we noticed, that array was wrong initially.
`defaultWindowCapability` map contains different capabilities for `MAIN` and for `WIDGET` windows.
As far as `["TEXT_WITH_GRAPHIC", "BUTTONS_WITH_GRAPHIC", "GRAPHIC_WITH_TEXT"]` is valid array for widgets, it is not valid for MAIN window which supports quite different templates.
That`s why we changed this array to reflect which templates are actually supported by MAIN window. Note, that we kept the same array unchanged for WIDGETS.

**Comment by AKalinich-Luxoft** [comment]
@kshala-ford all your questions have been addressed. Please let me know if you have additional questions.

File: app/model/sdl/Abstract/data.js (hunk 5 above)
Comment (kshala-ford) [reply app/model/sdl/Abstract/data.js:1028]: Not sure which templates this sdl_hmi supports for MAIN windows but the three templates `"TEXT_WITH_GRAPHIC", "BUTTONS_WITH_GRAPHIC", "GRAPHIC_WITH_TEXT"` are also possible templates for MAIN. They are not exclusively for widgets. Can you confirm that sdl_hmi  supports these three templates only for widgets?

**Comment by kshala-ford** [comment]
Thank you @AKalinich-Luxoft I wasn't aware so I retested two variables and yes they refer to the same object. Overall the proposal looks good to me. Just left a question for which templates are available.

File: app/model/sdl/Abstract/data.js (hunk 5 above)
Comment (AKalinich-Luxoft) [reply app/model/sdl/Abstract/data.js:1028]: @kshala-ford these three templates have no effect on the main window.
So how we checked that - just send `Show(templateConfiguration: {template: "<name>" })`. If you send one of these - `"MEDIA", "NON-MEDIA", "DEFAULT", "NAV_FULLSCREEN_MAP", 'WEB_VIEW'` then you may notice that HMI view template is changing for the main window and ignored for widgets.
If you send one of these - `"TEXT_WITH_GRAPHIC", "BUTTONS_WITH_GRAPHIC", "GRAPHIC_WITH_TEXT"`  then you may notice that HMI view template is NOT changing for the main window but it changes widgets view.
You definitely can send any template to any window (as this is just a string from API), however they will have no effect if it is not predefined.
I can confirm that currently we have 5 predefined templates for main window and 3 predefined templates for widgets.

**Comment by AKalinich-Luxoft** [comment]
> Thank you @AKalinich-Luxoft I wasn't aware so I retested two variables and yes they refer to the same object. Overall the proposal looks good to me. Just left a question for which templates are available.

@kshala-ford thank you for a quick review. I addressed your question reg templates

File: app/model/sdl/Abstract/data.js (hunk 5 above)
Comment (kshala-ford) [reply app/model/sdl/Abstract/data.js:1028]: Got it. Thank you!

**Comment by kshala-ford** [comment]
Ford approves the PR

**Comment by SilvioDC-Ford** [comment]
@theresalech this PR is ready for review and is approved by Ford.

**Comment by iCollin** [comment]
What will happen if the HMI attempts to deactivate an app while `templateChangeInProgress = true`?

Right now if I have two web view apps running and I open one of them will I see the iframe of both?

If I am viewing a WEB_VIEW app and I move to the CD player, the screen is blank. If I then click the phone tab, it seems the model is null:

```StateManager.js:493 Uncaught TypeError: Cannot read property 'appID' of null
    at d.exit (StateManager.js:493)
    at d [as exit] (ember-1.0.pre.min.js:14)```

**Comment by LitvinenkoIra** [comment]
@iCollin Thank you for the review, we are working on the reported issues

**Comment by YarikMamykin** [comment]
> What will happen if the HMI attempts to deactivate an app while `templateChangeInProgress = true`?
> 
> Right now if I have two web view apps running and I open one of them will I see the iframe of both?
> 
> If I am viewing a WEB_VIEW app and I move to the CD player, the screen is blank. If I then click the phone tab, it seems the model is null:
> 
> ```
> StateManager.js:493 Uncaught TypeError: Cannot read property 'appID' of null
>     at d.exit (StateManager.js:493)
>     at d [as exit] (ember-1.0.pre.min.js:14)
> ```

Hello @iCollin 
1. In case of app deactivation and `templateChangeInProgress = true` notification OnAppDeactivated() wouldn't be sent to SDL, 
but to catch such behavior you've got to be really fast :) 

2. We put minor fixes, so now you should see only one iframe relative to web app chosen per time.

3. Now you should be able to travel to CD player (Audio View) from web application view with no problems :)

File: app/model/sdl/Abstract/Model.js [hunk 6]
```diff
@@ -1496,7 +1496,9 @@ SDL.SDLModel = Em.Object.extend({
 
       SDL.TurnByTurnView.deactivate();
 
-      if (!SDL.SDLModel.data.phoneCallActive && reason == 'GENERAL') {
+      if (!SDL.SDLModel.data.phoneCallActive &&
+          !SDL.SDLModel.data.templateChangeInProgress &&
+          reason == 'GENERAL') {
         FFW.BasicCommunication.OnAppDeactivated(appID);
```
Comment (iCollin) [inline app/model/sdl/Abstract/Model.js:1502]: would it make sense to schedule this OnAppDeactivated to be send once the if condition is correct? I mean this deactivation shouldn't just be forgotten, correct?

Comment (AKalinich-Luxoft) [reply app/model/sdl/Abstract/Model.js:1502]: @iCollin yes, HMI should send this notification to SDL when it should be sent.
Flags, listed in this `if` statement are the conditions when HMI performs some kind of internal view switching without actual app deactivation. In that case notification should not be sent.
For example, when `templateChangeInProgress` is raised that means HMI is just deactivating current application template (like `WEB_VIEW`) and will activate a new template(like navi template) for this app so from user point of view this application is still active. That's why HMI should not send `OnAppDeactivated` in that case.

**Comment by iCollin** [comment]
@LuxoftAKutsan @AKalinich-Luxoft Hello, could you please fix the conflicts with develop so that this can be merged? Thank you!

**Comment by AKalinich-Luxoft** [comment]
> @LuxoftAKutsan @AKalinich-Luxoft Hello, could you please fix the conflicts with develop so that this can be merged? Thank you!

@iCollin done

**Pull request merged**